### PR TITLE
NCHWc: ReorderInput improvements

### PR DESCRIPTION
**Pull request by tracysh**
**Description**: 
Implement various improvements related to reordering a tensor for use by NCHWc operations:
1. Relax the requirement that the input channel count must be a multiple of the NCHWc block size (either 8 or 16 depending on ISA). The requirement now is that the channel count must be a multiple of 4. The implementation of MlasReorderInputNchw would need further work to support relaxing this further, but I don't have any models where I've observed this to be necessary yet.
2. Support fusing a Transpose(NHWC->NCHW) into a following ReorderInput. ReorderInput now has a channels_last attribute as was done in the past for ReorderOutput. This helps with models converted from TF where the converter is unable to remove all Transpose operations.
3. Add threading support to ReorderInput to accelerate performance (ReorderOutput will come later).

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/nchwc_ops.cc [hunk 1]
```diff
@@ -7,90 +7,128 @@
 namespace onnxruntime {
 namespace contrib {
 
-#define ONNX_CPU_OPERATOR_TYPED_NCHWC_KERNEL(name, ver, type, builder, ...) \
-  ONNX_OPERATOR_TYPED_KERNEL_EX(name, kMSNchwcDomain, ver, type, kCpuExecutionProvider, builder, __VA_ARGS__)
+Status ReorderInput::Compute(OpKernelContext* context) const {
+  const auto* X = context->Input<Tensor>(0);
+  const auto& X_shape = X->Shape().GetDims();
+  const auto X_rank = X_shape.size();
+  ORT_ENFORCE(X_rank >= 2);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/nchwc_ops.cc:14]: >2 [](http://example.com/codeflow?start=24&length=1)

do you mean 3?

File: onnxruntime/test/optimizer/nchwc_optimizer_test.cc [hunk 2]
```diff
@@ -1116,6 +1116,34 @@ TEST(NchwcOptimizerTests, BatchNormalization) {
...
+      EXPECT_EQ(op_to_count["Transpose"], 0);
+    };
+
+    // Verify that a NHWC->NCHW transpose is fused into ReorderInput.
+    NchwcOptimizerTester(build_test_case, check_nchwc_graph);
+  };
+
+  for (int64_t channels = 16; channels <= 32; channels += 4) {
+    test_case(channels);
+  }
```
Comment (yufenglee) [inline onnxruntime/test/optimizer/nchwc_optimizer_test.cc:1144]: could you add a case that channel is not mutiplier of 4

File: onnxruntime/contrib_ops/cpu/nchwc_ops.cc (hunk 1 above)
Comment (tracysh) [reply onnxruntime/contrib_ops/cpu/nchwc_ops.cc]: It was supposed to be any image tensor that has N+C, but I don't have tests to verify that, so I dropped it back to what it was before, checking that X_rank must be 4.

File: onnxruntime/test/optimizer/nchwc_optimizer_test.cc (hunk 2 above)
Comment (tracysh) [reply onnxruntime/test/optimizer/nchwc_optimizer_test.cc:1144]: Channels that are not a multiplier of 4 will fail. There is a test case for verifying the failure in TEST(NchwcOptimizerTests, TensorAlignment) (the first case covers this). Good enough?

Comment (yufenglee) [reply onnxruntime/test/optimizer/nchwc_optimizer_test.cc:1144]: Nice.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**